core/src/main/java/com/linecorp/armeria/client/metric/MetricCollectingClient.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.metric;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.function.Function;
+
+import com.linecorp.armeria.client.Client;
+import com.linecorp.armeria.client.ClientFactory;
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.SimpleDecoratingClient;
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.Response;
+import com.linecorp.armeria.common.metric.MetricKeyFunction;
+import com.linecorp.armeria.common.metric.Metrics;
+import com.linecorp.armeria.internal.metric.MetricCollectionSupport;
+
+/**
+ * Decorates a {@link Client} to collect metrics into {@link Metrics}.
+ * To use, build {@link Metrics}, specify it when building {@link ClientFactory} and
+ * add this decorator to a client.
+ *
+ * <p>Example:
+ * <pre>{@code
+ * MetricRegistry metricRegistry = new MetricRegistry();
+ * ClientFactory clientFactory = new ClientFactoryBuilder()
+ *         .metrics(new DropwizardMetrics(metricRegistry, "client"))
+ *         .build();
+ * MyService.Iface client = new ClientBuilder(uri)
+ *         .factory(clientFactory)
+ *         .decorate(MetricCollectingClient.newDecorator())
+ *         .build(MyService.Iface.class);
+ * }
+ * </pre>
+ *
+ * <p>It is generally recommended to define your own name for the service instead of using something like
+ * the Java class to make sure otherwise safe changes like renames don't break metrics.
```

Comment:
Rephrased so it's clearer:
```
 * <p>It is generally recommended not to use a class or package name as a metric name, because otherwise
 * seemingly harmless refactoring such as rename may break metric collection.
```